bluetooth/src/commonMain/kotlin/device/DeviceState.kt

```diff
@@ -0,0 +1,344 @@
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.launch
+import kotlin.coroutines.CoroutineContext
+
+sealed class DeviceAction {
+    sealed class Read : DeviceAction() {
+        class Characteristic(val characteristic: com.splendo.kaluga.bluetooth.Characteristic) : Read()
+        class Descriptor(val descriptor: com.splendo.kaluga.bluetooth.Descriptor) : Read()
+    }
+    sealed class Write(val newValue: ByteArray?) : DeviceAction() {
+        class Characteristic(newValue: ByteArray?, val characteristic: com.splendo.kaluga.bluetooth.Characteristic) : Write(newValue)
+        class Descriptor(newValue: ByteArray?, val descriptor: com.splendo.kaluga.bluetooth.Descriptor) : Write(newValue)
+    }
+
+    data class Notification(val characteristic: com.splendo.kaluga.bluetooth.Characteristic, val enable: Boolean) : DeviceAction()
+}
+
+typealias DeviceStateFlowRepo = StateRepo<DeviceState, MutableStateFlow<DeviceState>>
+
+sealed class DeviceState(
+    open val deviceInfo: DeviceInfoImpl,
+    internal open val connectionManager: BaseDeviceConnectionManager
+) : State(), DeviceInfo by deviceInfo, CoroutineScope by connectionManager {
+
+    sealed class Connected(
+        deviceInfo: DeviceInfoImpl,
+        connectionManager: BaseDeviceConnectionManager
+    ) : DeviceState(deviceInfo, connectionManager) {
+
+        data class NoServices internal constructor(
+            override val deviceInfo: DeviceInfoImpl,
+            override val connectionManager: BaseDeviceConnectionManager
+        ) : Connected(deviceInfo, connectionManager) {
+
+            fun startDiscovering() {
+                launch(coroutineContext) {
+                    connectionManager.stateRepo.takeAndChangeState { deviceState ->
+                        if (deviceState is NoServices)
+                            discoverServices
+                        else
+                            deviceState.remain()
+                    }
+                }
+            }
+
+            val discoverServices = suspend {
+                Discovering(deviceInfo, connectionManager)
+            }
+        }
+
+        data class Discovering internal constructor(
+            override val deviceInfo: DeviceInfoImpl,
+            override val connectionManager: BaseDeviceConnectionManager
+        ) : Connected(deviceInfo, connectionManager),
+            HandleAfterOldStateIsRemoved<DeviceState> {
+
+            internal fun didDiscoverServices(services: List<Service>): suspend () -> Idle {
+                return { Idle(services, deviceInfo, connectionManager) }
+            }
+
+            override suspend fun afterOldStateIsRemoved(oldState: DeviceState) {
+                connectionManager.discoverServices()
+            }
+        }
+
+        data class Idle internal constructor(
+            val services: List<Service>,
+            override val deviceInfo: DeviceInfoImpl,
```

Comment:
`deviceInfo` should be `DeviceInfo` interface here and not implementation

Reply:
The interface needs to be extended for this, but we should do that.